File: curvefit/nlinearfit.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import leastsq

#Create the x and y values
xdata = np.linspace(0, 10, 200)
ydata = 10 + 15 * np.exp(-xdata / 2) + np.random.randn(len(xdata))


#Define the exponential model
def my_fun(par, t, y):
    err = y - (par[0] + par[1] * np.exp(-t / par[2]))
    return err


#Function to evaluate the model
def fun_eval(par, t):
    return par[0] + par[1] * np.exp(-t / par[2])

#Initial guests
p0 = [min(ydata), max(ydata) - min(ydata), xdata[-3] / 3]

#Nonlinear fit
result = leastsq(my_fun, p0, args=(xdata, ydata))

#Plot the curves
plt.plot(xdata, ydata, 'k.')
plt.plot(xdata, fun_eval(result[0], xdata), 'r')
plt.xlabel("Time (s)")
plt.ylabel("Signal")
plt.show()

File: curvefit/nlinearfit.m
%Create x and y values
xdata = linspace(0, 10, 200);
ydata = 10 + 15.*exp(-xdata / 2) + randn(1, length(xdata));

%Define the exponential model
my_fun = @(p) sum((ydata- (p(1) + p(2).*exp(-xdata / p(3)))).^2);

%Function to evaluate the model. In this case we need to provide the error
%funtion: error sum of squares
fun_eval = @(p, t) p(1) + p(2).*exp(-t / p(3));

%Inital guesses
p0 = [min(ydata), max(ydata) - min(ydata), xdata(end) / 3]

%Nonlinear fit
result = fminsearch(my_fun, p0);

plot(xdata, ydata, 'k.')
hold on
plot(xdata, fun_eval(result, xdata), 'r')
xlabel('Time (s)')
ylabel('Signal')
